### PR TITLE
estimation implementation for both non gasless and gasless flow

### PR DESCRIPTION
**Pull request by talhamalik883**

### CODE REVIEW

File: src/AarcSDK.ts [hunk 1]
```diff
@@ -119,15 +126,36 @@ class AarcSDK {
...
+      const { receiverAddress, senderSigner } = executeMigrationDto;
+      let { transferTokenDetails } = executeMigrationDto;
       const owner = await senderSigner.getAddress();
-      const tokenAddresses = transferTokenDetails?.map((token) => token.tokenAddress);
+
+      const tokenAddresses = transferTokenDetails?.map(
+        (token) => token.tokenAddress,
+      );
+
+      if (tokenAddresses && tokenAddresses.length > 0) {
```
Comment (anshulforyou) [inline src/AarcSDK.ts:142]: What's the purpose of this code block?

File: src/helpers/PermitHelper.ts [hunk 2]
```diff
@@ -127,7 +126,7 @@ export class PermitHelper {
         ERC20_ABI,
         signer,
       );
-      const nonce = await tokenContract.nonces(eoaAddress);
+      const nonce = await this.getPermit2Nonce(eoaAddress);
```
Comment (anshulforyou) [inline src/helpers/PermitHelper.ts:129]: Here permit2 nonce will give wrong nonce. This function is signing msg for ERC20Permit contract. Here the nonce is calculated through the nonces function only. The previous code was right.

**Review by anshulforyou — CHANGES_REQUESTED** [review]
Please make sure that in gasless the ERC20Permit transaction is happening first with the  new flow as well.

File: tests/AarcSDK.test.ts [hunk 3]
```diff
@@ -5,272 +5,321 @@ import { AarcSDK } from '../src'; // Adjust the path according to your directory
...
+      tokenAddress: '0xf4ca1a280ebccdaebf80e3c128e55de01fabd893',
+      amount: expect.objectContaining({
+        _hex: '0x0f4240',
+      }),
+      message: 'Token transfer successful',
+      txHash: 'token-transfer-0x1234567890',
+    });
+  }, 30000);
+
+  it('should handle a successful migration for native token only', async () => {
```
Comment (anshulforyou) [inline tests/AarcSDK.test.ts:166]: Please create a different test suite file for native tokens

**Review by anshulforyou — CHANGES_REQUESTED** [review]

File: src/helpers/PermitHelper.ts (hunk 2 above)
Comment (talhamalik883) [reply src/helpers/PermitHelper.ts]: done

File: tests/AarcSDK.test.ts (hunk 3 above)
Comment (talhamalik883) [reply tests/AarcSDK.test.ts:211]: done

File: src/AarcSDK.ts [hunk 4]
```diff
@@ -7,6 +6,7 @@ import {
   PERMIT2_CONTRACT_ADDRESS,
   GELATO_RELAYER_ADDRESS,
   COVALENT_TOKEN_TYPES,
+  ETHEREUM_ADDRESS,
```
[the remark this answers is not in the document]
Comment (talhamalik883) [reply src/AarcSDK.ts]: done

File: src/AarcSDK.ts (hunk 1 above)
Comment (talhamalik883) [reply src/AarcSDK.ts:142]: If the user is going to supply token addresses, we will include native tokens to retrieve their balances as well. This way, we can verify if the native token holds enough value to facilitate the transfer of tokens.

**Review by anshulforyou — CHANGES_REQUESTED** [review]
We should add case in tests when the required gas is not there. What would be the response in that case?

**Review by anshulforyou — APPROVED** [review]

**Pull request merged**